include/kamping/distributed_graph_communicator.hpp

```diff
@@ -0,0 +1,353 @@
+            transform_elems_to_integers(in_edges, _in_weights.value(), get_weight);
+            transform_elems_to_integers(out_edges, _out_weights.value(), get_weight);
+        } else {
+            auto get_rank = [](auto const& edge) {
+                return static_cast<int>(edge);
+            };
+            transform_elems_to_integers(in_edges, _in_ranks, get_rank);
+            transform_elems_to_integers(out_edges, _out_ranks, get_rank);
+        }
+    }
+
+    /// @brief Constructs a communication graph based on a range of unweighted in-going and out-going edges.
+    CommunicationGraph(DefaultContainer<int>&& in_ranks, DefaultContainer<int>&& out_ranks)
+        : _in_ranks{std::move(in_ranks)},
+          _out_ranks{std::move(out_ranks)} {}
+
+    /// @brief Constructs a communication graph based on a range of weighted in-going and out-going edges.
+    CommunicationGraph(
+        DefaultContainer<int>&& in_ranks,
+        DefaultContainer<int>&& out_ranks,
+        DefaultContainer<int>&& in_weights,
+        DefaultContainer<int>&& out_weights
+    )
+        : _in_ranks{std::move(in_ranks)},
+          _out_ranks{std::move(out_ranks)},
+          _in_weights{std::move(in_weights)},
+          _out_weights{std::move(out_weights)} {}
+
+    /// @brief Constructs a communication graph based on a range of potentially weighted symmetric edges.
+    template <typename EdgeRange>
+    CommunicationGraph(EdgeRange const& edges) : CommunicationGraph(edges, edges) {}
+
+    /// @brief Returns a view of the communication graph.
+    CommunicationGraphView get_view() const {
+        if (_in_weights.has_value()) {
+            return CommunicationGraphView(_in_ranks, _out_ranks, _in_weights.value(), _out_weights.value());
+        } else {
+            return CommunicationGraphView(_in_ranks, _out_ranks);
+        }
+    }
+
+    /// @brief For each rank n for which there is (at least one) outgoing edge (own_rank, r), returns a mapping
+    /// r -> index of an edge (own_rank, r). This may be useful for collective communication primitives like
+    /// `MPI_Neighbor_alltoall` when one wants to calculate the position within the send_buffer to send data to a rank
+    /// r.
+    ///
+    /// @return Returns the mapping
+    template <typename Map = std::unordered_map<size_t, size_t>>
+    auto get_rank_to_out_edge_idx_mapping() {
+        Map mapping;
+        for (size_t i = 0; i < _out_ranks.size(); ++i) {
+            size_t rank = static_cast<size_t>(_out_ranks.data()[i]);
+            mapping.emplace(rank, i);
+        }
+        return mapping;
+    }
+
+private:
+    template <typename ReadFromRange, typename WriteToContainer, typename TransformOp>
+    void transform_elems_to_integers(ReadFromRange const& input, WriteToContainer& output, TransformOp&& transform_op) {
+        output.resize(input.size());
+        std::transform(input.data(), input.data() + static_cast<int>(input.size()), output.data(), transform_op);
```

Comment:
why cast this to `int` here? Also, in my opinion, extracting this to a separate function makes the code unnecessarily hard to read. Consider just inlining the `std::transform` call.

Reply:
I think it has to be signed, otherwise there will be a warning